### PR TITLE
Raylib5 bindings with fixes

### PR DESCRIPTION
**Pull request by perogycook**
I took seanarooni's pull request and fixed all the spacing issues and added more constants.  I couldn't contribute to that pull request so made my own fork and request, hope that's ok 😬.

### CODE REVIEW

**Comment by flysand7** [comment]
Based

**Comment by flysand7** [comment]
Please update raygui bindings to [4.0](https://github.com/raysan5/raygui/blob/master/src/raygui.h#L144) as well, these two libraries need to be in sync to work properly.

**Comment by perogycook** [comment]
> Please update raygui bindings to [4.0](https://github.com/raysan5/raygui/blob/master/src/raygui.h#L144) as well, these two libraries need to be in sync to work properly.

It already is at version 4.0 apparently, I added the missing definition you highlighted, the next newest changes all seem to be there...

**Comment by flysand7** [comment]
Hm interesting, there was my PR that added raygui bindings, and I remember compiling raygui 3.6 for linux, in fact it is 3.6 now that I look at it. Do the windows binaries provide 4.0?

Its good to know that foreign imports are correct, but the binaries need to be changed as well. lmk if you need the linux binaries, I can compile them on my machine and send them over or introduce them in a different PR

**Comment by perogycook** [comment]
> Hm interesting, there was my PR that added raygui bindings, and I remember compiling raygui 3.6 for linux, in fact it is 3.6 now that I look at it. Do the windows binaries provide 4.0?
> 
> Its good to know that foreign imports are correct, but the binaries need to be changed as well. lmk if you need the linux binaries, I can compile them on my machine and send them over or introduce them in a different PR

Weird, Linux lib at least seems to be 3.6 despite the commit being 4.0.  No idea, I don't have a Windows install.  Yeah if you can compile and send them that'd be great.  I was just fixing up stuff in codespace hah, I don't have Raylib source on my computer atm and it's almost 2am here.

**Comment by flysand7** [comment]
[raygui.zip](https://github.com/odin-lang/Odin/files/13466351/raygui.zip)

Here you go

**Comment by perogycook** [comment]
Thanks, added those.  Also, I hate Github lol.  Can't push from terminal with password smh.

**Comment by perogycook** [comment]
Looks like Windows n Mac most likely old too.  I'm traveling the next few days, don't have time to figure out cross compilation right now (not set up for it, just hammered out changes because I had a free hour in the evening)

**Comment by perogycook** [comment]
Little bump.  For those with a Mac or Windows install that want to help out, here's the link to Raygui:  https://github.com/raysan5/raygui

Windows instructions: 
```
copy src\raylib.h src\raylib.c
cl /O2 /I../raylib/src/ /D_USRDLL /D_WINDLL /DRAYGUI_IMPLEMENTATION /DBUILD_LIBTYPE_SHARED src/raygui.c /LD /Feraygui.dll /link /LIBPATH ../raylib/build/raylib/Release/raylib.lib /subsystem:windows /machine:x64
```

Mac instructions: 
```
cp src/raygui.h src/raygui.c
brew install raylib
gcc -o raygui.dynlib src/raygui.c -shared -fpic -DRAYGUI_IMPLEMENTATION -framework OpenGL -lm -lpthread -ldl $(pkg-config --libs --cflags raylib)
```

I've already compiled for Linux (and tested again to see exactly which libs are required since the instructions are rather sparce...), gist is that you need a C compiler, OpenGL headers and Raylib headers installed system-wide.

**Comment by karl-zylinski** [comment]
[raygui_windows.zip](https://github.com/odin-lang/Odin/files/13548857/raygui_windows.zip)

Here you go, built for windows. Hopefully it works for everyone 🤞

**Comment by laytan** [comment]
Here is the macos-arm64 build: 
[raygui-macos-arm64.zip](https://github.com/odin-lang/Odin/files/13549591/raygui-macos-arm64.zip)

**Comment by laytan** [comment]
Here is the macos-x86-64 build: 
[raygui-macos-x86-64.zip](https://github.com/odin-lang/Odin/files/13549767/raygui-macos-x86-64.zip)
NOTE: I built this using rosetta (emulation) from an arm machine and could not test it (I don't have an x86-64 mac machine) but I am pretty sure they will work.

File: vendor/raylib/raygui.odin [hunk 1]
```diff
@@ -34,7 +36,9 @@ when ODIN_OS == .Windows {
 		}
 	} else {
 		when RAYGUI_SHARED {
-			// #panic("Cannot link libraygui.450.dylib: not in the vendor collection")
+			foreign import lib {
+				"macos/libraygui.dylib"
```
Comment (karl-zylinski) [inline vendor/raylib/raygui.odin:40]: compile error: missing , at end of line

**Comment by karl-zylinski** [comment]
Ugh, there's a bug in Raylib. rl.IsMouseButtonUp is broken. It's fixed in the raylib master branch, but there's no patch release yet.

**Comment by karl-zylinski** [comment]
@perogycook I realized now that maybe we can hotfix this ourselves in the bindings by just making `IsMouseButtonUp` do `!IsMouseButtonDown()`

I.e. in the bindings change `IsMouseButtonUp       :: proc(button: MouseButton) -> bool ---`
to
`IsMouseButtonUp       :: proc(button: MouseButton) -> bool { return IsMouseButtonDown(button) }`

If you make that change then I can test it... I'll try to remember to undo this hotfix when an updated raylib comes out

**Comment by perogycook** [comment]
Ok done.

**Comment by karl-zylinski** [comment]
Hi! I tested your PR branch and it all works now except that I had to move the proc out of the foreign block, see patch below.

However, with this change everything works fine. If we just get these things fixed, then I would say this can be merged.

(can I somehow make PRs into your PR without having to ask you to edit stuff?)

```
diff --git a/vendor/raylib/raylib.odin b/vendor/raylib/raylib.odin
index d3632f002..830be6982 100644
--- a/vendor/raylib/raylib.odin
+++ b/vendor/raylib/raylib.odin
@@ -1189,10 +1189,10 @@ foreign lib {
 	IsMouseButtonPressed  :: proc(button: MouseButton) -> bool ---    // Detect if a mouse button has been pressed once
 	IsMouseButtonDown     :: proc(button: MouseButton) -> bool ---    // Detect if a mouse button is being pressed
 	IsMouseButtonReleased :: proc(button: MouseButton) -> bool ---    // Detect if a mouse button has been released once
+
+	// See IsMouseButtonUp below: This proc is broken is Raylib 5
 	// IsMouseButtonUp    :: proc(button: MouseButton) -> bool ---       Detect if a mouse button is NOT being pressed
-	IsMouseButtonUp       :: proc(button: MouseButton) -> bool {      // TODO: remove this when Raylib fixes this bug
-		return !IsMouseButtonDown(button)
-	}
+	
 	GetMouseX             :: proc() -> c.int ---                      // Returns mouse position X
 	GetMouseY             :: proc() -> c.int ---                      // Returns mouse position Y
 	GetMousePosition      :: proc() -> Vector2 ---                    // Returns mouse position XY
@@ -1709,7 +1709,10 @@ foreign lib {
 	DetachAudioMixedProcessor :: proc(processor: AudioCallback) --- // Detach audio stream processor from the entire audio pipeline
 }
 
-
+// TODO: remove this when Raylib releases a new binary version with this bug fixed
+IsMouseButtonUp :: proc(button: MouseButton) -> bool {
+	return !IsMouseButtonDown(button)
+}
 
 // Text formatting with variables (sprintf style)
 TextFormat :: proc(text: cstring, args: ..any) -> cstring { 
```

**Comment by perogycook** [comment]
Ok I added you as a collaborator to my repo, let me know if it works.  I agree, letting you just change things is easier hah, I'm just a newb when it comes to GitHub collaboration stuff.

**Comment by karl-zylinski** [comment]
@perogycook Thanks! I got it working and pushed the fix. I've the people in charge about this PR.

**Comment by DerTee** [comment]
> I.e. in the bindings change `IsMouseButtonUp :: proc(button: MouseButton) -> bool ---` to `IsMouseButtonUp :: proc(button: MouseButton) -> bool { return !IsMouseButtonDown(button) }`
> 
> If you make that change then I can test it... I'll try to remember to undo this hotfix when an updated raylib comes out


@karl-zylinski Is something like the following a viable solution to the remembering-to-undo-problem? It's not neat, but that's in the nature of workarounds:
```
// Raylib 5.0 has a bug in IsMouseButtonUp and this is a temporary workaround
when VERSION == "5.0" {
    `IsMouseButtonUp :: proc(button: MouseButton) -> bool { return !IsMouseButtonDown(button) }
} else {
    #panic("Remove the Raylib IsMouseButtonUp-Workaround that should be fixed in any Raylib version > 5.0")
    IsMouseButtonUp       :: proc(button: MouseButton) -> bool ---
}
```

**Comment by karl-zylinski** [comment]
@DerTee Thanks a good idea, thank you. I put something like that in (using two when blocks because the workaround must be outside the `foreign lib {` block).

I think this PR should be good to merge now @gingerBill, I've tested it on my game and it seems fine. The upgrade hassle for Raylib 4 -> 5 is minimal so I don't think we need to have two separate versions at this point.

**Pull request merged**